openhands/llm/llm.py

```diff
@@ -567,6 +567,7 @@ def format_messages_for_llm(self, messages: Message | list[Message]) -> list[dic
         for message in messages:
             message.cache_enabled = self.is_caching_prompt_active()
             message.vision_enabled = self.vision_is_active()
+            message.function_calling_enabled = self.is_function_calling_active()
```

Comment:
This will only be True for the selected set of model -- but we still send messages WITH tool_ids to the LLM

Reply:
Ahh so we need the part of the serializer that adds tool_call_id. OK, sure, I'll revert that. Sigh, this is messy. The attempt to separate things when function calling is native is because we need to send a single thing, not a list.

(I'll follow up on litellm too, they actually do this kind of conversion, so we don't have to play whack-a-mole anymore)

Reply:
@xingyaoww I actually restored this check, but it does send tool ids now. It just sends them as string, if non-native function calling, and as list if native. ([because this applies](https://github.com/All-Hands-AI/OpenHands/blob/07b96cc8c9507f963c9e971fdf8cceecc960fd71/openhands/llm/fn_call_converter.py#L456) now)

I started to work on refactoring the Message class. I think it needs to include tool calls/ids conversion part (at least as a single Message), do full serialization, and become closer to litellm's Message. I started to do it on this branch, but in the meantime, I found out that people rely on this PR, and use it via docker because it solves what's broken with non-native function calling.

So I'll make [another branch](https://github.com/All-Hands-AI/OpenHands/pull/5157) to work on the refactoring, and this PR is up for review again. If it's not too terrible, I'd say let's just fix the problems first, and refactor later.

(I'll clean up some leftover comment, too, but I prefer to do it if it's approved, because I think it would again break people's use via docker on the _last_ commit.)